Support computing strict transform of an ideal in Cox ring

paemurru opened the pull request:
Done.

paemurru committed:
Strict transform of an ideal in Cox ring

TODO: tests, docstrings, documentation

afkafkafk13 commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -44,3 +44,60 @@ end
+function _cox_ring_homomorphism(f::ToricBlowupMorphism)
+  Y = domain(f)
+  @assert grading_group(cox_ring(Y)) === class_group(Y)
+  G = class_group(Y)
```
First define G as the class_group(Y) and then use G in the assert, since you need G afterwards anyway and in the case of failure of the assert, you did not loose anything.

afkafkafk13 commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -44,3 +44,60 @@ end
+  n = number_of_generators(G)
+  new_var = cox_ring(Y)[index_of_new_ray(f)]
+  M = generator_degrees(cox_ring(Y))
+  @assert M[index_of_new_ray(f)][n] < 0 "Assuming the blowup adds a new column vector to `Oscar.generator_degrees(cox_ring(codomain(f)))`, the entry of that vector corresponding to the new ray should be negative. If not, multiply by -1."
```
Your error messages in case of failing asserts seem pretty long to me.  Please try to make them shorter and maybe add appropriate notes to the documentation instead.

afkafkafk13 commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -44,3 +44,60 @@ end
+  @assert M[index_of_new_ray(f)][n] < 0 "Assuming the blowup adds a new column vector to `Oscar.generator_degrees(cox_ring(codomain(f)))`, the entry of that vector corresponding to the new ray should be negative. If not, multiply by -1."
+  ind = Int64(abs(M[index_of_new_ray(f)][n]))
+  for i in 1:n_rays(Y)
+    i == index_of_new_ray(f) || @assert M[i][n] >= 0 "Assuming the blowup adds a new column vector to `Oscar.generator_degrees(cox_ring(codomain(f)))` for which the entry corresponding to the new ray is negative, the entries corresponding to all the other rays should be nonnegative. If not, add integer multiples of the other columns until it is."
```
Do you want to say 
`@assert findfirst(M[i][n] < 0 && i != index_of_new_ray(f) , 1:n_rays(Y)) === nothing`

afkafkafk13 reviewed (commented):
I leave you a few comments on the code. 

sideremark: I am not convinced (theorywise) that you can just set one entry to -1 in the grading change, but I am not a Cox-ring specialist. So I leave the reviewing of the content to someone else.

HechtiDerLachs commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -44,3 +44,60 @@ end
+
+function strict_transform(f::ToricBlowupMorphism, I::MPolyIdeal)
+  F, ind = _cox_ring_homomorphism(f)
+  Y = domain(f)
```
Thanks for looking into this! 

This is to compute the strict transform of an ideal in the cox ring of the codomain, correct? I would probably insert an `@assert` here, or something. 

Once this is done, I would be curious to compare performance with the chart-based approach!

afkafkafk13 commented:
@paemurru : Just a question, because this has been sitting here for 2 weeks now. What is the status here? Are you making progress on this? Are you waiting for comments from someone else? If so, please ping this person.

paemurru commented:
@afkafkafk13 I am working on this, this will probably stay as a draft pull request for quite some time. The current code only handles simple examples, the general case is a bit more complicated. The proof that it works uses a lot of notation, so is not suitable for a docstring or comment and must be typed in LaTeX. So it probably should be included in a research paper, unless an existing reference is found.

Essentially, we have an abelian group homomorphism, in general not a C-algebra homomorphism and in general not unique, between the Cox rings that takes homogeneous ideals to homogeneous ideals. And the image of a homogeneous ideal is its total transform.

paemurru committed:
Merge branch 'oscar-system:master' into ep/cox_ring_strict_transform

paemurru committed:
Partial progress

paemurru committed:
maximal proper subcones -> facets

paemurru committed:
Fix doctest

paemurru committed:
Progress on `_minimal_supercone`

paemurru committed:
Merge branch 'oscar-system:master' into ep/cox_ring_strict_transform

paemurru committed:
Merge branch 'oscar-system:master' into ep/cox_ring_strict_transform

paemurru committed:
Finish cox_ring_group_homomorphism for polynomials

Still need a corresponding function for ideals.

paemurru committed:
Fix tests

paemurru committed:
Fix according to JuliaFormatter

paemurru committed:
Add strict and total transform, add tests

paemurru committed:
Move sheaf-theoretic total transform in the file

To match the order of the documentation.

paemurru committed:
Fix unterminated doctest

paemurru committed:
Make tests clearer

paemurru committed:
Remove trailing whitespace

paemurru committed:
Simpler doctests

HechtiDerLachs commented on docs/src/AlgebraicGeometry/ToricVarieties/BlowupMorphisms.md:
````diff
@@ -79,7 +79,21 @@ morphisms are supported for toric blowups:
+strict_transform(f::ToricBlowupMorphism, I::MPolyIdeal)
+strict_transform_with_index(f::ToricBlowupMorphism, I::MPolyIdeal)
+```
+The above functions are implemented using a group homomorphism between the Cox rings that takes homogeneous ideals to homogeneous ideals:
````
I'm slightly confused: Is it intentional that we speak of a "group homomorphism" here? I would expect a ring homomorphism. Maybe this is because the ring homomorphism is induced from some morphism of lattices/groups?

paemurru replied:
https://github.com/oscar-system/Oscar.jl/pull/4154#discussion_r1873528692

HechtiDerLachs commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+julia> R = cox_ring(X)
+Multivariate polynomial ring in 2 variables over QQ graded by
+  x1 -> 1 by 0 empty matrix
+  x2 -> 1 by 0 empty matrix
```
This seems odd. Sorry to bring this up, since this is probably not the main point of this PR, but does everyone agree that this is what we want here?

paemurru replied:
Made the pull request https://github.com/oscar-system/Oscar.jl/pull/4383

HechtiDerLachs commented on experimental/Schemes/src/ToricBlowups/methods.jl:
````diff
@@ -1,3 +1,231 @@
+```
+"""
+function total_transform(f::ToricBlowupMorphism, I::MPolyIdeal)
+  return cox_ring_group_homomorphism(f, I)
````
The example above suggests that the return value of this is indeed an ideal. Looking at the code, my naive guess would have been that we get a homomorphism whose construction depends on `f` and `I`. But it seems that in fact we apply a homomorphism associated to `f` to `I`. To make this a bit more streamlined with existing functionality, may I suggest to make the current implementation for `cox_ring_homomorphism` for this signature a method for `pullback(f, I)`? Unless this signature is ambiguous, it should be more or less what we want here, right?

paemurru replied:
It is not actually a ring homomorphism, only a homomorphism of modules. Also see https://github.com/oscar-system/Oscar.jl/pull/4154#discussion_r1873526484

HechtiDerLachs commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+  S = cox_ring(domain(f))
+  new_var = S[index_of_new_ray(f)]
+  J = total_transform(f, I)
+  return saturation(J, ideal(new_var))
```
```suggestion
  new_var = S[index_of_new_ray(f)]
  J = total_transform(f, I)
  return saturation(J, ideal(S, new_var))
```
Just a nitpick, but to avoid potential future problems.

HechtiDerLachs commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+This is implemented under the following assumptions:
+  * the star subdivision adds a ray (instead of subdividing along an existing ray),
+  * $X$ is an orbifold (meaning its fan is simplicial), and
+  * $X$ has no torus factors (meaning the rays span $N_{\mathbb{R}}$).
```
I think it's great that you worked out the mathematics behind this, thank you for that! However, on a first glance the implementation below does not seem to have checks for these requirements. Is this because there is no algorithm for these checks? 

We do have some of these cases; for instance there is no feasible way to assure that a covered scheme is proper over a point. But unless this is the case here, I suggest to introduce the common optional `check::Bool=true` argument in the function's signature and some internal checks which make sure that these three things hold true.

paemurru replied:
All of the polyhedral geometry checks will be much faster than saturation, so I think it is OK to always do them

HechtiDerLachs commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+end
+
+@doc raw"""
+    strict_transform_with_index(f::ToricBlowupMorphism, I::MPolyIdeal) -> MPolyIdeal
```
```suggestion
    strict_transform_with_index(f::ToricBlowupMorphism, I::MPolyIdeal) -> (MPolyIdeal, Int)
```

HechtiDerLachs commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+    cox_ring_group_homomorphism(f::ToricBlowupMorphism, I::MPolyIdeal) -> MPolyIdeal
+
+Let $f\colon Y \to X$ be the toric blowup corresponding to a star subdivision along a ray. Let $R$ and $S$ be the Cox rings of $X$ and $Y$, respectively.
+Considering $R$ and $S$ with their abelian group structures, there is a group homomorphism $\Phi\colon R \to S$ sending homogeneous ideals to homogeneous ideals with the defining property that the the total transform of a subscheme of $X$, given by a homogeneous ideal $I$ in $R$, is the subscheme of $Y$ given by the homogeneous ideal $\Phi(I)$ in $S$.
```
```suggestion
Considering $R$ and $S$ with their abelian group structures, there is a group homomorphism $\Phi\colon R \to S$ sending homogeneous ideals to homogeneous ideals with the defining property that the total transform of a subscheme of $X$, given by a homogeneous ideal $I$ in $R$, is the subscheme of $Y$ given by the homogeneous ideal $\Phi(I)$ in $S$.
```

HechtiDerLachs commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+"""
+cox_ring_group_homomorphism
+
+function cox_ring_group_homomorphism(f::ToricBlowupMorphism, g::MPolyDecRingElem)
```
Make these methods of `pullback(f::ToricBlowupMorphism, ...)`? See my above comment.

paemurru replied:
Since it is not actually a pullback and does not satisfy some expected properties of pullback (since the homomorphism is not generally a ring homomorphism), I would rather not call it pullback

afkafkafk13 replied:
I understand that this is not a pullback, but you do not return a group homomorphism either. Can you come up with any way to clearly phrase the name of this method so that users are not misled by the name? Is there a specific name for the returned object in literature?

paemurru replied:
My initial thought was to actually return a group homomorphism `Phi`, and then let the user call `Phi(g)`. And store the group homomorphism as an attribute. But storing the homomorphism had some difficulties with function closures. So I instead turned it into taking two arguments, `f` and `g`

HechtiDerLachs commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+  Y = domain(f)
+  @req is_normal(X) "Only implemented when the variety is normal"
+  @req !has_torusfactor(X) "Only implemented when there are no torus factors"
+  @req n_rays(Y) == n_rays(X) + 1 "Only implemented when the blowup adds a ray"
```
Referring to one of my above comments: It seems that here the sanity of the input is checked. Maybe these checks are then also triggered by the above code. Either way: It looks like some of these tests could be rather expensive, so I would like to point out the option to use `@check` instead of `@req` here so that turning of these tests internally is possible. What do you think?

paemurru replied:
Only the `has_torusfactor` check is expensive. Related pull request: https://github.com/oscar-system/Oscar.jl/pull/4385

paemurru replied:
https://github.com/oscar-system/Oscar.jl/pull/4154#discussion_r1884136347

HechtiDerLachs commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+  n = nvars(cox_ring(X))
+  phi_exponents = zeros(ZZ, n)
+  for i in 1:n
+    ray = Oscar._ray_variable_correspondence(X)[cox_ring(X)[i]]
```
```suggestion
    ray = _ray_variable_correspondence(X)[cox_ring(X)[i]]
```

HechtiDerLachs commented on src/PolyhedralGeometry/PolyhedralFan/properties.jl:
```diff
@@ -281,6 +281,49 @@ function cones(PF::_FanLikeType)
 
+@doc raw"""
+    minimal_supercone(X::NormalToricVariety, p::AbstractVector{<:IntegerUnion})
+    -> safksjafld
```
```suggestion
    -> ???
```

paemurru replied:
Fixed in [ff5df80](https://github.com/oscar-system/Oscar.jl/pull/4154/commits/ff5df80f61c91bdc1790dbd913388526fad0685e)

HechtiDerLachs commented on src/PolyhedralGeometry/PolyhedralFan/properties.jl:
```diff
@@ -281,6 +281,49 @@ function cones(PF::_FanLikeType)
+  m_fan = polyhedral_fan(m_cone)
+
+  # By Proposition 1.2.8(c) in [CLS11](@cite), every proper face of a cone
+  # is the intersection of the facets containing it.
```
I think comments like this are great; thank you for that!

HechtiDerLachs reviewed (commented):
Thank you for your work here @paemurru. As I understand, these things have not yet been written up or implemented properly and it looks like you took care of working out the details. That said, I have to admit that I am probably not in the position to check all the mathematical details of this implementation, but I feel that this is also why we have people like you on board: You probably know best what's correct here and I would like to trust you with this. 

From what I can see, the contributions to the code base seem to fit with the existing code. I left some comments to indicate potential for streamlining a bit more. Let me know what you think.

paemurru committed:
Update experimental/Schemes/src/ToricBlowups/methods.jl

Co-authored-by: Matthias Zach <85350711+HechtiDerLachs@users.noreply.github.com>

paemurru committed:
Update experimental/Schemes/src/ToricBlowups/methods.jl

Co-authored-by: Matthias Zach <85350711+HechtiDerLachs@users.noreply.github.com>

paemurru committed:
Fix temporary name

paemurru committed:
Update experimental/Schemes/src/ToricBlowups/methods.jl

Co-authored-by: Matthias Zach <85350711+HechtiDerLachs@users.noreply.github.com>

paemurru committed:
Update experimental/Schemes/src/ToricBlowups/methods.jl

Co-authored-by: Matthias Zach <85350711+HechtiDerLachs@users.noreply.github.com>

HereAround commented on src/PolyhedralGeometry/PolyhedralFan/standard_constructions.jl:
```diff
@@ -204,7 +204,7 @@ function _check_containment_via_facet_signs(smaller::Vector{Int}, bigger::Vector
 end
 function _get_maximal_cones_containing_vector(
-  Sigma::_FanLikeType, v::AbstractVector{<:IntegerUnion}
+  Sigma::_FanLikeType, v::AbstractVector{<:Union{Integer,ZZRingElem,QQFieldElem}}
```
I do not yet understand why `QQFieldElem` is needed here. Can you explain? Also, maybe write this as union of `integerUnion` and `QQFieldElem`, so that future updates to IntegerUnion affect/arrive at this line.

paemurru replied:
Changed to `RationalUnion`

HereAround commented on src/PolyhedralGeometry/PolyhedralFan/properties.jl:
```diff
@@ -281,6 +281,49 @@ function cones(PF::_FanLikeType)
 end
 
+@doc raw"""
+    minimal_supercone(X::NormalToricVariety, p::AbstractVector{<:IntegerUnion})
```
This is a method in polyhedral fans, so the input should be a polyhedral fan. Right?

If you allow for polyhedral fans as input, then they can have lineality. Right? Where is that case treated in your code?

Final question: Why not just return a cone? Why is an iterator needed/better here? At least from the name of the method, I expect a cone as output.

HereAround replied:
If you really want a toric variety as first argument, then probably this method should be moved to the toric code.

paemurru replied:
Moved to PolyhedralGeometry

HereAround commented on test/PolyhedralGeometry/polyhedral_fan.jl:
```diff
@@ -119,6 +119,36 @@
   end
 end
 
+@testset "minimal_supercone" begin
```
The following tests should be moved to the toric tests, as they test toric stuff. (Or you accept a polyhedral fan as first argument.)

paemurru replied:
Moved to PolyhedralGeometry

HereAround reviewed (changes requested):
Some of the code in this PR is about toric varieties, but in the polyhedral folder. Two options:
1. Allow polyhedral fans as input and leave the code where it is.
2. Move the code to the toric varieties code/tests.

@lkastner will eyeball the `minimal_supercone` method. As currently is, this is likely not as efficient/performant as it could be.

lkastner commented on src/PolyhedralGeometry/PolyhedralFan/properties.jl:
```diff
@@ -281,6 +281,49 @@ function cones(PF::_FanLikeType)
+  cs = [c for c in m_facets if r in c]
+  isempty(cs) && return m_cone
+  return intersect(cs)
+end
```
This is too expensive. Why are you intersecting cones that belong to the same fan? You should intersect the corresponding index sets. Why construct a fan out of `m_cone`? The intersection of cones `m_cone` is already the minimal cone containing `r`. I am lost for the remainder of this function. My suggestion is the following:

```suggestion
function minimal_supercone(
  X::NormalToricVariety, r::AbstractVector{<:Union{IntegerUnion,QQFieldElem}}
)
  m_cone_indices = Oscar._get_maximal_cones_containing_vector(X, r)
  @req !is_empty(m_cone_indices) "Ray `r` should be in the support of the fan of `X`."
  IX = maximal_cones(IncidenceMatrix, X)
  m_cone = intersect([row(IX, i) for i in m_cone_indices]...)
  RX = matrix(coefficient_field(X), rays(X))
  return positive_hull(RX[[m_cone...], :], lineality_space(X))
end
```

Probably @benlorenz has further improvements.

paemurru replied:
> The intersection of cones m_cone is already the minimal cone containing r

No. If the fan has one maximal cone, then `m_cone` is just that one cone, which is not necessarily the minimal cone in the fan containing the point. In the attached picture, the point `(1, 0, 1)` is in the relative interior of a two-dimensional cone in the fan. That is the cone that this function returns.

![Screenshot from 2024-12-06 15-13-45](https://github.com/user-attachments/assets/521ddb70-e48d-4206-b41d-991b69d3e162)

paemurru replied:
It would be good if we can only work with indices and perhaps just return the index of the cone. I was not sure how to do this

lkastner replied:
You're right. My claim is even wrong when there are more maximal cones. I apologize for that.

lkastner replied:
I will work something out using the Hasse diagram.

lkastner replied:
In the end this is the same as your approach, but I managed to avoid expensive intersections:
```
julia> m_cone = positive_hull([1 0 0; 0 1 0; 0 0 1])
Polyhedral cone in ambient dimension 3

julia> r = [1,1, 0]
3-element Vector{Int64}:
 1
 1
 0

julia> r in m_cone
true

julia> Fm = facets(m_cone)
3-element SubObjectIterator{LinearHalfspace{QQFieldElem}} over the halfspaces of R^3 described by:
-x_1 <= 0
-x_2 <= 0
-x_3 <= 0


julia> zero_facets = findall(f -> f.a*r == [0], Fm)
1-element Vector{Int64}:
 3

julia> Rm = matrix(coefficient_field(m_cone), rays(m_cone))
[1   0   0]
[0   1   0]
[0   0   1]

julia> RIF = IncidenceMatrix(m_cone.pm_cone.RAYS_IN_FACETS)
3×3 IncidenceMatrix
[2, 3]
[1, 3]
[1, 2]


julia> result_ray_indices = intersect([row(RIF, i) for i in zero_facets]...)
Set{Int64} with 2 elements:
  2
  1

julia> result = positive_hull(Rm[[result_ray_indices...], :], lineality_space(m_cone))
Polyhedral cone in ambient dimension 3
```
So the general thing could look like:
```suggestion
function minimal_supercone(
  X::NormalToricVariety, r::AbstractVector{<:Union{IntegerUnion,QQFieldElem}}
)
  m_cone_indices = Oscar._get_maximal_cones_containing_vector(X, r)
  @req !is_empty(m_cone_indices) "Ray `r` should be in the support of the fan of `X`."
  IX = maximal_cones(IncidenceMatrix, X)
  m_cone = intersect([row(IX, i) for i in m_cone_indices]...)
  RX = matrix(coefficient_field(X), rays(X))
  m_cone = positive_hull(RX[[m_cone...], :], lineality_space(X))
  Fm = facets(m_cone)
  zero_facets = findall(f -> f.a*r == [0], Fm)
  Rm = matrix(coefficient_field(m_cone), rays(m_cone))
  RIF = IncidenceMatrix(m_cone.pm_cone.RAYS_IN_FACETS)
  result_ray_indices = intersect([row(RIF, i) for i in zero_facets]...)
  result = positive_hull(Rm[[result_ray_indices...], :], lineality_space(m_cone))
  return result
end
```
This avoids the expensive intersections, nevertheless it requires a convex hull computation. Might be more if the fan did not have rays before. Hope this works for you.

paemurru replied:
The tests failed when some of the defined sets were empty. I pushed a commit that dealt with three such special cases in the above. Now the tests seem to pass, so I hope it is correct.

Returning the index of the resulting cone would also be OK if that is faster.

One side-comment that I am less familiar with polymake, so the variable names RIF, Fm, RX, IX, etc are very hard to read for me.

paemurru replied:
Instead of tests with affine_space and projective_space, I guess the tests can be replaced by fan tests, so it can be entirely under PolyhedralGeometry

lkastner commented on src/exports.jl:
```diff
@@ -1068,6 +1068,7 @@ export minimal_block_reps
 export minimal_faces
 export minimal_generating_set
 export minimal_size_generating_set, has_minimal_size_generating_set, set_minimal_size_generating_set
+export minimal_supercone
```
Do we need to export this?

paemurru replied:
@simonbrandhorst recommended to me to export it since the code is documented, general and reasonably fast

lkastner commented on src/PolyhedralGeometry/PolyhedralFan/properties.jl:
```diff
@@ -281,6 +281,49 @@ function cones(PF::_FanLikeType)
 
+@doc raw"""
+    minimal_supercone(X::NormalToricVariety, p::AbstractVector{<:IntegerUnion})
+    -> SubObjectIterator{RayVector}
```
```suggestion
```

paemurru replied:
Why not show the output type?

lkastner commented on src/PolyhedralGeometry/PolyhedralFan/properties.jl:
```diff
@@ -281,6 +281,49 @@ function cones(PF::_FanLikeType)
+    minimal_supercone(X::NormalToricVariety, p::AbstractVector{<:IntegerUnion})
+    -> SubObjectIterator{RayVector}
+
+Given an point $p$ with integer coordinates inside the support of the fan $Σ$ of a normal toric variety~$X$, return the unique cone $σ$ in $Σ$ such that $p$ is in the relative interior of $σ$.
```
Please wrap

lkastner commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+    cox_ring_group_homomorphism(f::ToricBlowupMorphism, I::MPolyIdeal) -> MPolyIdeal
+
+Let $f\colon Y \to X$ be the toric blowup corresponding to a star subdivision along a ray. Let $R$ and $S$ be the Cox rings of $X$ and $Y$, respectively.
+Considering $R$ and $S$ with their abelian group structures, there is a group homomorphism $\Phi\colon R \to S$ sending homogeneous ideals to homogeneous ideals with the defining property that the total transform of a subscheme of $X$, given by a homogeneous ideal $I$ in $R$, is the subscheme of $Y$ given by the homogeneous ideal $\Phi(I)$ in $S$.
```
Please wrap

lkastner commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+Returns the pair $(J, k)$, where $J$ coincides with `strict_transform(f, I)` and where $k$ is the multiplicity of the total transform along the exceptional prime divisor.
+
+!!! note
+    If the multiplicity $k$ is not needed, we recommend to use `strict_transform(f, I)` which is typically faster.
```
Please wrap

lkastner commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+
+Let $f\colon Y \to X$ be the toric blowup corresponding to a star subdivision along a ray. Let $R$ and $S$ be the Cox rings of $X$ and $Y$, respectively.
+Here "strict transform" means the "scheme-theoretic closure of the complement of the exceptional divisor in the scheme-theoretic inverse image".
+This function returns a homogeneous ideal in $S$ corresponding to the strict transform under $f$ of the closed subscheme of $X$ defined by the homogeneous ideal $I$ in $R$.
```
Please wrap

lkastner commented on experimental/Schemes/src/ToricBlowups/methods.jl:
```diff
@@ -1,3 +1,231 @@
+    total_transform(f::ToricBlowupMorphism, I::MPolyIdeal) -> MPolyIdeal
+
+Let $f\colon Y \to X$ be the toric blowup corresponding to a star subdivision along a ray. Let $R$ and $S$ be the Cox rings of $X$ and $Y$, respectively.
+This function returns a homogeneous ideal in $S$ corresponding to the total transform (meaning the scheme-theoretic inverse image) under $f$ of the closed subscheme of $X$ defined by the homogeneous ideal $I$ in $R$.
```
Please wrap

lkastner reviewed (changes requested):
Please update the PR description. Are these TODOs done now?

paemurru committed:
Merge branch 'oscar-system:master' into ep/cox_ring_strict_transform

paemurru committed:
Tests do not depend on order of rays anymore

paemurru committed:
Fix trivial grading doctests

paemurru committed:
Wrap comments

paemurru committed:
Fix output type

lgoettgens commented on test/PolyhedralGeometry/polyhedral_fan.jl:
```diff
@@ -119,6 +119,36 @@
+  X = projective_space(NormalToricVariety, 3)
+  c = Oscar.minimal_supercone(X, [1, 1, 1])
+  @test rays(c) isa SubObjectIterator{RayVector{QQFieldElem}}
+  @test sort(collect(rays(c))) == sort([[1, 0, 0], [0, 1, 0], [0, 0, 1]])
```
```suggestion
  @test issetequal(rays(c)), [[1, 0, 0], [0, 1, 0], [0, 0, 1]])
```
and similar for the other tests below. Sorting needs resources that can be saved this way

paemurru replied:
I don't think it works

```
julia>   X = projective_space(NormalToricVariety, 3)
Normal toric variety

julia>   c = Oscar.minimal_supercone(X, [1, 1, 1])
Polyhedral cone in ambient dimension 3

julia> sort(rays(c)) == sort([[1, 0, 0], [0, 1, 0], [0, 0, 1]])
true

julia> issetequal(rays(c), [[1, 0, 0], [0, 1, 0], [0, 0, 1]])
false
```

lgoettgens replied:
Hm, weird. I thought this was how @benlorenz adapted the other tests recently.

paemurru replied:
Perhaps the difference of types explains the output?

```
julia> typeof(rays(c))
SubObjectIterator{RayVector{QQFieldElem}}

julia> typeof([[1, 0, 0], [0, 1, 0], [0, 0, 1]])
Vector{Vector{Int64}}
```

benlorenz replied:
`issetequal` wont work if the hash-functions are not compatible which is not really feasible when comparing different types with different semantics.
This should work:
```suggestion
  @test issetequal(rays(c)), ray_vector.([[1, 0, 0], [0, 1, 0], [0, 0, 1]]))
```

paemurru replied:
Updated the tests to use `issetequal` as @benlorenz suggested.

> which is not really feasible

It looks feasible to me:

```
julia> issetequal(Int64[1, 2], Int8[1, 2])
true

julia> issetequal(Int64[1, 2], ZZRingElem[1, 2])
true
```

It is not for this pull request, but perhaps the hash function of QQFieldElem should be altered.

```
julia> issetequal(Int64[1, 2], QQFieldElem[1, 2])
false
```

paemurru committed:
Update test/PolyhedralGeometry/polyhedral_fan.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

paemurru committed:
Undo issetequal

paemurru committed:
Update src/PolyhedralGeometry/PolyhedralFan/properties.jl

Co-authored-by: Lars Kastner <lkastner@users.noreply.github.com>

paemurru committed:
Fix cases where list is empty

paemurru committed:
JuliaFormatter

paemurru committed:
Merge branch 'oscar-system:master' into ep/cox_ring_strict_transform

paemurru committed:
Polyhedral geometry tests now use issetequal

paemurru committed:
Move some code to PolyhedralGeometry

Also: add more tests, move `@req` from `cox_ring_module_homomorphism` to `strict_transform`, `total_transform` and `strict_transform_with_index`.

Changed `minimal_supercone` to `minimal_supercone_indices`, which is what is needed for `minimal_supercone_coordinates`, which is what I actually need for strict transforms.

Everything should be fixed now and should work properly.

paemurru committed:
Fix documentation

paemurru committed:
JuliaFormatter

paemurru committed:
Add function `primitive_generator`

Also fix `minimal_supercone_coordinates` and add alias
`minimal_generator`.

paemurru committed:
Fix `strict_transform` for nonsimplicial

Add test for nonsimplicial, add comment to
`minimal_supercone_coordinates` for the nonsimplicial case.

paemurru committed:
Fix docs

Also fix `minimal_supercone_coordinates_of_new_ray`

paemurru committed:
Merge branch 'oscar-system:master' into ep/cox_ring_strict_transform

lkastner commented on src/PolyhedralGeometry/PolyhedralFan/properties.jl:
```diff
@@ -281,6 +281,163 @@ function cones(PF::_FanLikeType)
+"""
+function primitive_generator(r::RayVector)
+  return vec(collect(matrix(ZZ, rays(positive_hull(r)))))
+end
```
```suggestion
function primitive_generator(r::RayVector)
  result = numerator.(lcm([denominator(i) for i in r]) * r)
  result = 1/gcd(result) * result
  convert(Vector{ZZRingElem}, result)
end
```

lkastner replied:
`positive_hull` and then asking for `rays` triggers a convex hull computation.
```
julia> t
3-element RayVector{QQFieldElem}:
 1
 5//4
 3//2

julia> @benchmark vec(collect(matrix(ZZ, rays(positive_hull(t)))))
BenchmarkTools.Trial: 242 samples with 1 evaluation.
 Range (min … max):  19.107 ms … 35.539 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     20.441 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   20.674 ms ±  1.326 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

          ▁ ▁██▁                                               
  ▂▂▄▄▄▄▄▅█▇████▇▆▅▅▅▄▃▃▄▃▄▃▁▁▁▁▁▁▂▂▂▁▁▁▁▁▁▁▁▂▁▁▂▁▁▁▁▁▁▁▁▂▂▁▂ ▃
  19.1 ms         Histogram: frequency by time        25.6 ms <

 Memory estimate: 10.81 KiB, allocs estimate: 932.

julia> @benchmark convert(Vector{ZZRingElem}, 1/gcd(numerator.(lcm([denominator(i) for i in t]) * t)) * (numerator.(lcm([denominator(i) for i in t]) * t)))
BenchmarkTools.Trial: 10000 samples with 8 evaluations.
 Range (min … max):  3.585 μs … 105.043 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     4.280 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   4.788 μs ±   1.813 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▄ ▁█▆▃▅▆▄▃▂▂▁▁▁           ▁▁                                ▂
  ██████████████████▇▆▇▆▇█▇███▇▇▇▇▇▆▅▅▅▃▄▁▁▄▄▃▄▁▁▁▃▁▄▁▁▃▁▁▁▄▇ █
  3.58 μs      Histogram: log(frequency) by time      13.7 μs <

 Memory estimate: 2.88 KiB, allocs estimate: 107.

julia> convert(Vector{ZZRingElem}, 1/gcd(numerator.(lcm([denominator(i) for i in t]) * t)) * (numerator.(lcm([denominator(i) for i in t]) * t))) == vec(collect(matrix(ZZ, rays(positive_hull(t)))))
true
```

paemurru replied:
Great improvement, thanks!

paemurru committed:
Update src/PolyhedralGeometry/PolyhedralFan/properties.jl

Co-authored-by: Lars Kastner <lkastner@users.noreply.github.com>

paemurru committed:
Fix according to JuliaFormatter

HereAround reviewed (approved):
Thank you for your hard work on this PR @paemurru and @lkastner . Looks great!
